era-yul/src/yul/visitor/mod.rs

```diff
@@ -0,0 +1,159 @@
+//!
+//! Implementation of a visitor pattern for YUL syntax tree.
+//!
+
+use std::collections::BTreeSet;
+
+use crate::yul::parser::statement::assignment::Assignment;
+use crate::yul::parser::statement::block::Block;
+use crate::yul::parser::statement::code::Code;
+use crate::yul::parser::statement::expression::function_call::name::Name;
+use crate::yul::parser::statement::expression::function_call::FunctionCall;
+use crate::yul::parser::statement::expression::literal::Literal;
+use crate::yul::parser::statement::expression::Expression;
+use crate::yul::parser::statement::for_loop::ForLoop;
+use crate::yul::parser::statement::function_definition::FunctionDefinition;
+use crate::yul::parser::statement::if_conditional::IfConditional;
+use crate::yul::parser::statement::object::Object;
+use crate::yul::parser::statement::switch::Switch;
+use crate::yul::parser::statement::variable_declaration::VariableDeclaration;
+use crate::yul::parser::statement::Statement;
+
+use super::parser::dialect::Dialect;
+
+///
+/// Utility conventional name of a function corresponding to the `code` block of
+/// an object.
+///
+pub const IMPLICIT_CODE_FUNCTION_NAME: &str = "BODY";
+
+///
+/// Create a virtual definition of a function corresponding to the `code` block
+/// of an object.
+///
+pub fn implicit_code_function<P>(code: &Code<P>) -> FunctionDefinition<P>
+where
+    P: Dialect,
+{
+    FunctionDefinition {
+        location: code.location,
+        identifier: IMPLICIT_CODE_FUNCTION_NAME.to_string(),
+        arguments: Vec::new(),
+        result: Vec::new(),
+        body: code.block.clone(),
+        attributes: BTreeSet::new(),
+    }
+}
+
+#[allow(unused_variables)]
+///
+/// Visitor for YUL syntax tree.
+///
+pub trait Visitor<P>
+where
+    P: Dialect,
+{
+    ///
+    /// Visit `switch` statement in YUL syntax tree.
+    ///
+    fn visit_switch(&mut self, switch: &Switch<P>) {
+        unreachable!("{}", Self::MSG_METHOD_NOT_IMPLEMENTED)
```

Comment:
All these methods are supposed to be optional, aren't they?

Reply:
My idea was that all these methods are optional to define, but if they are used, the implementation has to be provided. Therefore I've made default implementations panic. What would you suggest to change?

Reply:
Nah it's fine. Just confirming.